### PR TITLE
Ensure Vector3 inequality doesn't use the hidden 4th element

### LINKED ISSUE #36905: System.Numerics.Tests.Vector3Tests.Vector3InequalityTest fails on arm64
### Builds
|Build|Pull Request | Test Failure Count|
| --- | --- | --- |
|[#656665](https://dev.azure.com/dnceng/public/_build/results?buildId=656665)|Rolling|2|
|[#656760](https://dev.azure.com/dnceng/public/_build/results?buildId=656760)|Rolling|2|
|[#656870](https://dev.azure.com/dnceng/public/_build/results?buildId=656870)|Rolling|2|
|[#657104](https://dev.azure.com/dnceng/public/_build/results?buildId=657104)|Rolling|2|
|[#657257](https://dev.azure.com/dnceng/public/_build/results?buildId=657257)|Rolling|2|
### Configurations
- net5.0-Linux-Release-arm64-CoreCLR_release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-16.04-helix-arm64v8-bfcd90a-20200127194925
- net5.0-Windows_NT-Release-arm64-CoreCLR_release-Windows.10.Arm64.Open
### Helix Logs
|Build|Pull Request|Console|Core|Test Results|Run Client|
| --- | --- | --- | --- | --- | --- |
|[#656665](https://dev.azure.com/dnceng/public/_build/results?buildId=656665)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/7f71b271-da78-4e13-b176-25f4c498c7cc/workitems/System.Numerics.Vectors.Tests/files/console.5d9596da.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/7f71b271-da78-4e13-b176-25f4c498c7cc/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-7f71b271da784e13b1/System.Numerics.Vectors.Tests/f58f62da-c3a8-485e-a986-f228b51ccbdd.log?sv=2019-02-02&se=2020-06-11T04%3A43%3A51Z&sr=c&sp=rl&sig=a9U4H9GhhebdXpHGTHqnxshUcdC8JPS6W38aU4QWD%2Fc%3D)|
|[#656665](https://dev.azure.com/dnceng/public/_build/results?buildId=656665)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/39230b6e-4fca-4538-95cd-3bb29971e408/workitems/System.Numerics.Vectors.Tests/files/console.772c5426.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/39230b6e-4fca-4538-95cd-3bb29971e408/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-39230b6e4fca453895/System.Numerics.Vectors.Tests/9c7021a8-273d-4d69-80f5-7a5bb40e27f3.log?sv=2019-02-02&se=2020-06-11T04%3A57%3A48Z&sr=c&sp=rl&sig=q9IIfKhX65kYULTCu64fWB1IXEHJJREdls6BVWEnz7o%3D)|
|[#656760](https://dev.azure.com/dnceng/public/_build/results?buildId=656760)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/9e1f2845-7643-4003-931f-ac56a09d2f55/workitems/System.Numerics.Vectors.Tests/files/console.7d4c6737.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/9e1f2845-7643-4003-931f-ac56a09d2f55/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-9e1f28457643400393/System.Numerics.Vectors.Tests/63fb8dbe-3bd8-44f6-a3cb-46a21e872a77.log?sv=2019-02-02&se=2020-06-11T07%3A18%3A20Z&sr=c&sp=rl&sig=aAWBuN%2FTKg5XpbmpkcWYetMFta8Vg6Pm6vDDo31veNU%3D)|
|[#656760](https://dev.azure.com/dnceng/public/_build/results?buildId=656760)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/df8f323e-6099-4458-aa99-b102bbe0c184/workitems/System.Numerics.Vectors.Tests/files/console.0db17546.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/df8f323e-6099-4458-aa99-b102bbe0c184/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-df8f323e60994458aa/System.Numerics.Vectors.Tests/f37ca826-bcc0-4cb3-95d7-b12106a9bd6c.log?sv=2019-02-02&se=2020-06-11T07%3A41%3A06Z&sr=c&sp=rl&sig=%2BrLAp2Hy58c2MO4NUptnoSSK3otbLO8F9PSxib6Knf4%3D)|
|[#656870](https://dev.azure.com/dnceng/public/_build/results?buildId=656870)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/d551ab26-a301-4236-a2aa-a71e02c936fc/workitems/System.Numerics.Vectors.Tests/files/console.5bdc24b4.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/d551ab26-a301-4236-a2aa-a71e02c936fc/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-d551ab26a3014236a2/System.Numerics.Vectors.Tests/0d3914c6-b83e-4812-92bf-c60223d31c3e.log?sv=2019-02-02&se=2020-06-11T09%3A09%3A59Z&sr=c&sp=rl&sig=IxQm8m7bhXo0GY%2F54TDQFM1J8jIkXNbjdwoKrl3vCU4%3D)|
|[#656870](https://dev.azure.com/dnceng/public/_build/results?buildId=656870)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/e1664939-492b-4606-9e8a-a1864685cc76/workitems/System.Numerics.Vectors.Tests/files/console.7ae51d4b.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/e1664939-492b-4606-9e8a-a1864685cc76/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-e1664939492b46069e/System.Numerics.Vectors.Tests/d71da648-6a63-48ef-9933-333adc636f7f.log?sv=2019-02-02&se=2020-06-11T09%3A26%3A36Z&sr=c&sp=rl&sig=lnDtnlxOC0uTrU4ZCCLhRM5IK%2BITa6IBNjaGbyDQ3KQ%3D)|
|[#657104](https://dev.azure.com/dnceng/public/_build/results?buildId=657104)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/c11fc99c-efdf-4065-ac26-02163855b218/workitems/System.Numerics.Vectors.Tests/files/console.355fe2b5.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/c11fc99c-efdf-4065-ac26-02163855b218/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-c11fc99cefdf4065ac/System.Numerics.Vectors.Tests/bda77534-934e-4e28-83b6-f06a2a2a441e.log?sv=2019-02-02&se=2020-06-11T13%3A57%3A06Z&sr=c&sp=rl&sig=6LHKGfUq80P3mkPIG5om6Y0j7aW7c4GydU5pIFVhkUs%3D)|
|[#657104](https://dev.azure.com/dnceng/public/_build/results?buildId=657104)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/ee5c1dd4-3b14-4fe5-96ca-bcb5403ce094/workitems/System.Numerics.Vectors.Tests/files/console.60c14362.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/ee5c1dd4-3b14-4fe5-96ca-bcb5403ce094/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-ee5c1dd43b144fe596/System.Numerics.Vectors.Tests/48b74c28-94d0-4ac9-82e5-da734f191318.log?sv=2019-02-02&se=2020-06-11T14%3A14%3A44Z&sr=c&sp=rl&sig=Q95y6Un0XkIpwApANm4Rgg0r3XbOZrZYjknjkt%2BVPjw%3D)|
|[#657257](https://dev.azure.com/dnceng/public/_build/results?buildId=657257)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/d5c6bc31-bc1e-48c5-bb17-6910f9c8ccd6/workitems/System.Numerics.Vectors.Tests/files/console.8b0bb602.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/d5c6bc31-bc1e-48c5-bb17-6910f9c8ccd6/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-d5c6bc31bc1e48c5bb/System.Numerics.Vectors.Tests/67b17911-d719-4720-822e-b1a18a3542ec.log?sv=2019-02-02&se=2020-06-11T17%3A29%3A06Z&sr=c&sp=rl&sig=hYfPNchmVUFFOGqJ9gtkRYQO5J9VpNI1DfoqjcHjrBA%3D)|
|[#657257](https://dev.azure.com/dnceng/public/_build/results?buildId=657257)|Rolling|[console.log](https://helix.dot.net/api/2019-06-17/jobs/95894e9e-7cec-417e-a31c-dba116ae2877/workitems/System.Numerics.Vectors.Tests/files/console.2e48f2ec.log)||[testResults.xml](https://helix.dot.net/api/2019-06-17/jobs/95894e9e-7cec-417e-a31c-dba116ae2877/workitems/System.Numerics.Vectors.Tests/files/testResults.xml)|[run_client.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-95894e9e7cec417ea3/System.Numerics.Vectors.Tests/abbce9de-3c2c-4dcb-806c-8fd83672cefe.log?sv=2019-02-02&se=2020-06-11T17%3A40%3A36Z&sr=c&sp=rl&sig=QjUvqffQyC%2BTl4tllQvng%2BUZnBYnugqMgbMfkMKb6HA%3D)|

```
  Discovering: System.Numerics.Vectors.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Numerics.Vectors.Tests (found 1194 test cases)
  Starting:    System.Numerics.Vectors.Tests (parallel test collections = on, max threads = 4)
    System.Numerics.Tests.Vector3Tests.Vector3InequalityTest [FAIL]
      Assert.Equal() Failure
      Expected: False
      Actual:   True
      Stack Trace:
        /_/src/libraries/System.Numerics.Vectors/tests/Vector3Tests.cs(978,0): at System.Numerics.Tests.Vector3Tests.Vector3InequalityTest()
  Finished:    System.Numerics.Vectors.Tests
```

Note at the moment arm64 tests are only part of rolling builds due to hardware limitation.

### PR DESCRIPTION
**Pull request by tannergooding**
This fixes #36905

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @safern, @jkotas, @CarolEidt

File: src/coreclr/src/jit/lowerarmarch.cpp [hunk 1]
```diff
@@ -706,12 +706,18 @@ void Lowering::LowerHWIntrinsicCmpOp(GenTreeHWIntrinsic* node, genTreeOps cmpOp)
...
     if ((baseType == TYP_FLOAT) && (simdSize == 12))
     {
-        // For TYP_SIMD12 we need to clear the upper bits and can't assume their value
+        // For TYP_SIMD12 the value of the upper bits is unknown but shouldn't impact
+        // the comparison. So, we should always pretend the bits were equivalent to
+        // ensure that the final result is correct. In the case of equality, the unsigned
+        // min across will always treat -1 (all bits set) as higher and so it will be
+        // preserved if the other three elements were also equal. In the case of inequality
+        // it will likewise be handled and won't cause an otherwise equivalent vector from
+        // being treated as not equal.
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lowerarmarch.cpp:715]: I found this just a bit confusing. I might reword as
```suggestion
        // For TYP_SIMD12 we don't want the upper bits to participate in the comparison. So, we will insert all ones
        // into those bits of the result, "as if" the upper bits are equal. Then if all lower bits are equal, we get the 
        // expected all-ones result, and will get the expected 0's only where there are non-matching bits.
```
Or something along those lines.

**Review by CarolEidt — APPROVED** [review]
LGTM - just one optional comment suggestion.

**Review by safern — APPROVED** [review]
Did you do a rolling build from your branch to make sure the test passes on all arm test runs? 

FWIW, I'm enabling arm64 libraries testing using a checked runtime on coreclr changes: https://github.com/dotnet/runtime/pull/36910

File: src/coreclr/src/jit/lowerarmarch.cpp [hunk 2]
```diff
@@ -706,12 +706,14 @@ void Lowering::LowerHWIntrinsicCmpOp(GenTreeHWIntrinsic* node, genTreeOps cmpOp)
 
     if ((baseType == TYP_FLOAT) && (simdSize == 12))
     {
-        // For TYP_SIMD12 we need to clear the upper bits and can't assume their value
+        // For TYP_SIMD12 we don't want the upper bits to participate in the comparison. So, we will insert all ones
+        // into those bits of the result, "as if" the upper bits are equal. Then if all lower bits are equal, we get the 
+        // expected all-ones result, and will get the expected 0's only where there are non-matching bits.
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lowerarmarch.cpp:711]: Thanks for making the comment change - it probably overflowed the allowed line length causing the formatting error? Sorry about that!

**Comment by tannergooding** [comment]
@safern, no, I added a regression test to the CLR side instead to ensure this gets tested irrespective of the libraries tests running.

**Pull request merged**